### PR TITLE
many: handle core configuration internally instead of using the core configure hook

### PR DESCRIPTION
**Pull request by mvo5**
A early preview mostly to validate directions.

### CODE REVIEW

File: overlord/configstate/configmgr.go [hunk 1]
```diff
@@ -22,23 +22,106 @@ package configstate
...
+			return err
+		}
+	} else {
+		if err := t.Get("patch", &patch); err != nil && err != state.ErrNoState {
+			return err
+		}
+	}
+
+	tr := config.NewTransaction(st)
+	for key, value := range patch {
```
Comment (pedronis) [inline overlord/configstate/configmgr.go]: transaction don't need the state lock until commit afaiu/r, so we could unlock from here until after runCoreCfg

File: overlord/configstate/configmgr.go [hunk 2]
```diff
@@ -22,23 +22,106 @@ package configstate
...
+	defer st.Unlock()
+
+	// FIXME: duplicated code from configureHandler.Before()
+	if err := t.Get("use-defaults", &useDefaults); err != nil && err != state.ErrNoState {
+		println(err.Error())
+		return err
+	}
+	if useDefaults {
+		var err error
+		patch, err = snapstate.ConfigDefaults(st, snapName)
```
Comment (pedronis) [inline overlord/configstate/configmgr.go]: this could just hard code "core", as we do below

File: overlord/configstate/configmgr.go [hunk 3]
```diff
@@ -22,23 +22,106 @@ package configstate
...
+	var useDefaults bool
+
+	st := t.State()
+	snapName := "core"
+
+	// FIXME: lock taken for too long
+	st.Lock()
+	defer st.Unlock()
+
+	// FIXME: duplicated code from configureHandler.Before()
```
Comment (pedronis) [inline overlord/configstate/configmgr.go]: helper then?

File: overlord/configstate/configmgr.go [hunk 4]
```diff
@@ -22,23 +22,106 @@ package configstate
...
+	}
+
+	tr := config.NewTransaction(st)
+	for key, value := range patch {
+		if err := tr.Set("core", key, value); err != nil {
+			return nil
+		}
+	}
+
+	if err := corecfgRun(tr); err != nil {
```
Comment (pedronis) [inline overlord/configstate/configmgr.go]: we will need to pass State as well at some point, or add a State method to transaction as some other things have

File: overlord/configstate/configmgr.go [hunk 5]
```diff
@@ -22,23 +22,106 @@ package configstate
...
+	// snap is installed.
+	runner := state.NewTaskRunner(st)
 	manager := &ConfigManager{
-		state: s,
+		state:  st,
+		runner: runner,
 	}
-
-	hookManager.Register(regexp.MustCompile("^configure$"), newConfigureHandler)
+	runner.AddHandler("run-core-configure", manager.doRunCoreConfigure, nil)
```
Comment (pedronis) [inline overlord/configstate/configmgr.go]: Gustavo suggestion for this was "configure-snapd"

File: corecfg/corecfg.go [hunk 6]
```diff
@@ -22,59 +22,51 @@ package corecfg
...
+type Conf interface {
+	Get(snapName, key string, result interface{}) error
 }
 
-func snapctlGet(key string) (string, error) {
-	raw, err := exec.Command("snapctl", "get", key).CombinedOutput()
-	if err != nil {
-		return "", fmt.Errorf("cannot run snapctl: %s", osutil.OutputErr(raw, err))
-	}
+func snapctlGet(tr Conf, key string) (string, error) {
```
Comment (pedronis) [inline corecfg/corecfg.go]: coreGet would probably be a better name for this now

**Comment by pedronis** [comment]
this:

	// see if it makes sense to run at all
	if release.OnClassic {
		return fmt.Errorf("cannot run core-configure on classic distribution")
	}

is too blunt now, and we are hitting all the time, we need to decide a different approach, like only error on specific core only keys on classic

**Comment by pedronis** [comment]
btw, can be done later but now we can move refresh.schedule syntax checks here as well

File: overlord/configstate/configmgr.go (hunk 4 above)
Comment (mvo5) [reply overlord/configstate/configmgr.go]: I added a State() method to transaction now.

**Pull request merged**